### PR TITLE
Feature/random test set

### PR DESCRIPTION
**Pull request by jan-gerling**
### Fix Bugs:
 * fixed incorrect output formatting for test scores
 * fix minor bug in QueryBuilder test_get_level_refactorings_count

### Features:
 * multiple validation sets for the model evaluation
    * specify validation sets in config
 * move DROP_PROCESS_AND_AUTHORSHIP_METRICS column specification to logs
 * adopt output to generate logs for multiple validation sets
 * every call to log also prints

### Other
* Refactor preprocessing retrieve_labelled_instances

### CODE REVIEW

File: machine-learning/ml/pipelines/binary.py [hunk 1]
```diff
@@ -39,48 +65,82 @@ def run(self):
...
+
+                    x_tests, y_tests, dataset_name = [], [], []
+                    for validation_dataset in VALIDATION_DATASETS:
+                        dataset_name.append(validation_dataset)
+                        test_features, x_test, y_test, _ = retrieve_labelled_instances(validation_dataset, refactoring, False, train_features)
+                        # test if any refactorings were found for the given refactoring type
+                        if x_test is None:
+                            log("Skip model building for refactoring type: " + refactoring.name())
+                            continue
+                        if len(x_tests) == 0:
```
Comment (mauricioaniche) [inline machine-learning/ml/pipelines/binary.py:90]: Do you need this if? Can't you use the `append` directly?

File: machine-learning/ml/pipelines/binary.py [hunk 2]
```diff
@@ -39,48 +65,82 @@ def run(self):
...
+                            x_tests, y_tests = [x_test], [y_test]
+                        else:
+                            x_tests.append(x_test)
+                            y_tests.append(y_test)
+
+                    # shuffle X and Y again, just to be save
+                    x = pd.concat([x_train] + x_tests).sample(frac=1, random_state = 42)
```
Comment (mauricioaniche) [inline machine-learning/ml/pipelines/binary.py:97]: **IMPORTANT**: You can't shuffle them separately. Otherwise, `x` and `y` won't match anymore!!!!

Comment (mauricioaniche) [reply machine-learning/ml/pipelines/binary.py:97]: You would have to put them in a complete data frame, shuffle, and then, break them again in x and y

Comment (mauricioaniche) [reply machine-learning/ml/pipelines/binary.py:97]: (I have done this mistake before, that's why I spotted it!)

File: machine-learning/ml/pipelines/binary.py [hunk 3]
```diff
@@ -94,47 +154,18 @@ def _run_single_model(self, model_def, x, y, x_train, x_test, y_train, y_test):
...
-    def _build_production_model(self, model_def, best_params, x, y):
-        log("Production model build started at %s\n" % now())
-
-        super_model = model_def.model(best_params)
-        super_model.fit(x, y)
-
-        return super_model
-
-
 class DeepLearningBinaryClassificationPipeline(BinaryClassificationPipeline):
```
Comment (mauricioaniche) [inline machine-learning/ml/pipelines/binary.py:166]: I think we can delete everything related to deep learning? Or are we going to try some NN?

File: machine-learning/ml/pipelines/binary.py [hunk 4]
```diff
@@ -1,18 +1,46 @@
...
+def _build_production_model(model_def, best_params, x, y):
+    log("Production model build started at %s\n" % now())
+
+    super_model = model_def.model(best_params)
+    super_model.fit(x, y)
+
+    return super_model
+
+
+def _evaluate_model(search, x_train, x_tests, y_train, y_tests):
```
[the remark this answers is not in the document]
Comment (mauricioaniche) [reply machine-learning/ml/pipelines/binary.py:22]: I see now, so indeed, we will have three sets [github, validation 1, validation 2]. We always train on GitHub and use others to validate. When we validate in validation 1, we "lose" validation 2 data points. 

Am I correct?

File: machine-learning/ml/pipelines/binary.py [hunk 5]
```diff
@@ -39,48 +65,82 @@ def run(self):
...
-                        log("An error occurred while working on refactoring " + refactoring_name + " model " + model.name())
-                        log(str(e))
-                        log(str(traceback.format_exc()))
-
-
-    def _run_single_model(self, model_def, x, y, x_train, x_test, y_train, y_test):
+                # we have two options to select a test set,
+                # 1.) Predefined in the database
+                if TEST_SPLIT_SIZE < 0 and len(VALIDATION_DATASETS) > 0:
+                    train_features, x_train, y_train, scaler = retrieve_labelled_instances(dataset, refactoring, True)
```
Comment (mauricioaniche) [inline machine-learning/ml/pipelines/binary.py:76]: should we extract the content of this `if` and then the `else`, and put in two separated methods? 

`train_test_split_per_project` and `train_test_split_per_percentage` or something like that?

File: machine-learning/configs.py [hunk 6]
```diff
@@ -70,7 +75,11 @@
 # Evaluation: Cross-validation configuration
 # --------------------------------
 
-TEST_SPLIT_SIZE = 0.2
+# Specify either a train/ test split, e.g. 0.2 -> 80/ 20 split
+TEST_SPLIT_SIZE = -1
+# Or specify test data sets in the database
+# NOTE: set TEST_SPLIT_SIZE value to < 0, in order to indicate to use the given datasets instead of a random train/ test split
+VALIDATION_DATASETS = ["test set github", "validation set github"]
```
[the remark this answers is not in the document]
Comment (jan-gerling) [reply machine-learning/configs.py:82]: Yes

File: machine-learning/utils/log.py [hunk 7]
```diff
@@ -50,7 +50,9 @@ def log_close():
     _f.close()
 
 
-def log(msg):
+def log(msg, print_msg: bool = True):
```
[the remark this answers is not in the document]
Comment (jan-gerling) [reply machine-learning/utils/log.py:53]: Yes, I got remove quite some duplicate print/ log statements with this tweak

File: machine-learning/ml/pipelines/binary.py [hunk 8]
```diff
@@ -39,48 +65,82 @@ def run(self):
         for dataset in self._datasets:
             log("Dataset {}".format(dataset))
 
+            # test set github, validation set github
```
[the remark this answers is not in the document]
Comment (jan-gerling) [reply machine-learning/ml/pipelines/binary.py:68]: Yes, I should change this.

File: machine-learning/ml/pipelines/binary.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (jan-gerling) [reply machine-learning/ml/pipelines/binary.py:22]: We always train on the Github dataset and use all (2) validation sets to generate the scores for the model, by iterating over all validation sets in evaluate_model. Thus, we do not "lose" any validation sets.

File: machine-learning/ml/pipelines/binary.py (hunk 1 above)
Comment (jan-gerling) [reply machine-learning/ml/pipelines/binary.py:90]: I will try it

File: machine-learning/ml/pipelines/binary.py (hunk 3 above)
Comment (jan-gerling) [reply machine-learning/ml/pipelines/binary.py:166]: I don't think, we have the time for NN or other deep learning algorithms.
I will remove it in another PR though, to keep this one more concise.

Comment (jan-gerling) [reply machine-learning/ml/pipelines/binary.py:166]: https://github.com/refactoring-ai/predicting-refactoring-ml/issues/201

File: machine-learning/ml/pipelines/binary.py [hunk 9]
```diff
@@ -39,48 +65,82 @@ def run(self):
...
+                    train_features, x_train, y_train, scaler = retrieve_labelled_instances(dataset, refactoring, True)
+                    # test if any refactorings were found for the given refactoring type
+                    if x_train is None:
+                        log("Skip model building for refactoring type: " + refactoring.name())
+                        continue
+
+                    x_tests, y_tests, dataset_name = [], [], []
+                    for validation_dataset in VALIDATION_DATASETS:
+                        dataset_name.append(validation_dataset)
+                        test_features, x_test, y_test, _ = retrieve_labelled_instances(validation_dataset, refactoring, False, train_features)
```
[the remark this answers is not in the document]
Comment (jan-gerling) [reply machine-learning/ml/pipelines/binary.py:85]: Done

File: machine-learning/ml/pipelines/binary.py (hunk 2 above)
Comment (jan-gerling) [reply machine-learning/ml/pipelines/binary.py:97]: Fixed this

File: machine-learning/ml/pipelines/binary.py (hunk 1 above)
Comment (jan-gerling) [reply machine-learning/ml/pipelines/binary.py:90]: Works

**Comment by jan-gerling** [comment]
> 1. You can't shuffle `x` and `y` separately.
> 2. We have to use the scaler generated in the training to build the validation dataset. Otherwise, things won't match.

Both major issues are fixed.

File: machine-learning/ml/pipelines/binary.py (hunk 5 above)
Comment (jan-gerling) [reply machine-learning/ml/pipelines/binary.py:75]: I don't see the benefit, there is little duplicate code and the pipeline would not get any easier.

File: machine-learning/ml/pipelines/binary.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (mauricioaniche) [reply machine-learning/ml/pipelines/binary.py:22]: but we miss training points, right?

if we are validating using set 1, then the data points of validation set 2 could be used for training. I guess that's not a problem due to the amount of the data we have anyways.

File: machine-learning/ml/preprocessing/preprocessing.py [hunk 10]
```diff
@@ -63,38 +48,41 @@ def retrieve_labelled_instances(dataset, refactoring: LowLevelRefactoring):
...
         assert x.shape[0] == y.shape[0], "Balancing did not work, x and y have different shapes."
-        log("instances after balancing: {}".format(Counter(y)))
+        log("instances after balancing: {}".format(Counter(y)), False)
 
     # apply some scaling to speed up the algorithm
-    scaler = None
-    if SCALE_DATASET:
+    if SCALE_DATASET and scaler is None:
         x, scaler = perform_scaling(x)
+    elif SCALE_DATASET and scaler is not None:
```
Comment (mauricioaniche) [inline machine-learning/ml/preprocessing/preprocessing.py:76]: Do we need to replace the scaler here? I'm supposing it returns the same one, right?

File: machine-learning/ml/preprocessing/scaling.py [hunk 11]
```diff
@@ -13,6 +13,21 @@ def perform_scaling(x):
     scaler = MinMaxScaler()  # Default behavior is to scale to [0,1]
     columns = x.columns
     x = scaler.fit_transform(x)
-    x = pd.DataFrame(x, columns=columns) # keeping the column names
+    x = pd.DataFrame(x, columns=columns)  # keeping the column names
 
     return x, scaler
+
+
+def perform_scaling(x, scaler):
```
Comment (mauricioaniche) [inline machine-learning/ml/preprocessing/scaling.py:21]: maybe this method should not return the scaler, as it uses it but does not produce it.

File: machine-learning/ml/pipelines/binary.py (hunk 1 above)
Comment (mauricioaniche) [reply machine-learning/ml/pipelines/binary.py:90]: But I still see the if here

Can't it just be

```
x_tests.append(x_test)
y_tests.append(y_test)
```

without the if?

Comment (jan-gerling) [reply machine-learning/ml/pipelines/binary.py]: It got last in the merge, I fixed that.

File: machine-learning/ml/preprocessing/scaling.py (hunk 11 above)
Comment (jan-gerling) [reply machine-learning/ml/preprocessing/scaling.py:21]: Changed it

File: machine-learning/ml/pipelines/binary.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (jan-gerling) [reply machine-learning/ml/pipelines/binary.py:23]: We don't use them for the training of the initial model, which we also evaluate, but we use the validation to train the final (super) model anyways. Also, due to the enormous amount of training data, that should not be an issue at all.
